clash-prelude/src/Clash/Signal/Bundle.hs

```diff
@@ -166,3 +170,34 @@ instance Bundle ((f :*: g) a) where
    where
     getL (l :*: _) = l
     getR (_ :*: r) = r
+
+-- | Internal data type, see "TaggedUnit".
+data TaggedUnit' (dom :: k) = TaggedUnit
+
+-- | Helper type to emulate the "old" behavior of Bundle's unit instance. I.e.,
+-- the instance for @Bundle ()@ used to be defined as:
+--
+--     class Bundle () where
+--       bundle   :: () -> Signal domain ()
+--       unbundle :: Signal domain () -> ()
+--
+-- In order to have sensible type inference, the "Bundle" class specifies that
+-- the argument type of 'bundle' should uniquely identify the result type, and
+-- vice versa for 'unbundle'. The type signatures in the snippet above don't
+-- though, as @()@ doesn't uniquely map to a specific domain. In other words,
+-- 'domain' should occur in both the argument and result of both functions.
+--
+-- "TaggedUnit" fixes this by carrying the domain in its type. Use 'TaggedUnit'
+-- to construct a "TaggedUnit" or a "TaggedUnit'".
+type TaggedUnit (dom :: Domain) = TaggedUnit' '(dom, dom)
```

Comment:
It actually occurs to me now that I don't so much like the naming here. I think that `TaggedUnit` should be the name for the general type (instead of `TaggedUnit'`) and the specific tags should have a name which is more separate.

I also have reservations about how conceptually similar this is to `(:::)`, essentially being `(::: ())`. Is there some middle ground where maybe that type is changed / newtyped to provide this behaviour?

Reply:
As discussed offline: I don't want users to write `TaggedUnit' '(dom, dom)`. But I agree it's kinda meh.

Reply:
w.r.t. the `(:::)` becoming a newtype: they’re type-aliases to be very lightweight, making them a newtype negates that. Unless, we create a constraint-solver plugin that automatically (unsafe)Coerces from/to `(:::)`. But I wouldn’t want to spend much time on such a plug-in now.